src/MongoDB.Driver.Core/Core/WireProtocol/Messages/Encoders/BinaryEncoders/CompressedMessageBinaryEncoder.cs

```diff
@@ -0,0 +1,151 @@
+    /// </summary>
+    public sealed class CompressedMessageBinaryEncoder : MessageBinaryEncoderBase, IMessageEncoder
+    {
+        private readonly ICompressorSource _compressorSource;
+        private readonly MessageEncoderSettings _encoderSettings;
+        private readonly IMessageEncoderSelector _originalEncoderSelector;
+        private const int MessageHeaderLength = 16;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CompressedMessageBinaryEncoder" /> class.
+        /// </summary>
+        /// <param name="stream">The stream.</param>
+        /// <param name="originalEncoderSelector">The original encoder selector.</param>
+        /// <param name="compressorSource">The compressor source.</param>
+        /// <param name="encoderSettings">The encoder settings.</param>
+        public CompressedMessageBinaryEncoder(
+            Stream stream,
+            IMessageEncoderSelector originalEncoderSelector,
+            ICompressorSource compressorSource,
+            MessageEncoderSettings encoderSettings)
+            : base(stream, encoderSettings)
+        {
+            _compressorSource = Ensure.IsNotNull(compressorSource, nameof(compressorSource));
+            _encoderSettings = encoderSettings; // can be null
+            _originalEncoderSelector = originalEncoderSelector;
+        }
+
+        /// <summary>
+        /// Reads the message.
+        /// </summary>
+        /// <returns>A message.</returns>
+        public CompressedMessage ReadMessage()
+        {
+            var reader = CreateBinaryReader();
+            var stream = reader.BsonStream;
+
+            var messageStartPosition = stream.Position;
+            var messageLength = stream.ReadInt32();
+            EnsureMessageLengthIsValid(messageLength);
+            var requestId = stream.ReadInt32();
+            var responseTo = stream.ReadInt32();
+            var opcode = (Opcode)stream.ReadInt32();
+            EnsureOpcodeIsValid(opcode);
+            var originalOpcode = (Opcode)stream.ReadInt32();
+            var uncompressedSize = stream.ReadInt32();
+            var compressorType = (CompressorType)stream.ReadByte();
+            var compressor = _compressorSource.Get(compressorType);
+
+            using (var uncompressedBuffer = new MultiChunkBuffer(new OutputBufferChunkSource(BsonChunkPool.Default)))
+            using (var uncompressedStream = new ByteBufferStream(uncompressedBuffer, ownsBuffer: false))
+            {
+                uncompressedStream.WriteInt32(uncompressedSize + MessageHeaderLength);
+                uncompressedStream.WriteInt32(requestId);
+                uncompressedStream.WriteInt32(responseTo);
+                uncompressedStream.WriteInt32((int)originalOpcode);
+                compressor.Decompress(stream, uncompressedStream);
+                uncompressedStream.Position = 0;
+                uncompressedBuffer.MakeReadOnly();
```

Comment:
Small fix to be able to use `originalMessageEncoder`

Reply:
Good catch.

I would find a blank line between the setting up of the uncompressed buffer and the following lines of code helpful.